Tests/MongoSwiftTests/BSONValueTests.swift

```diff
@@ -152,4 +153,64 @@ final class BSONValueTests: MongoSwiftTestCase {
         expect(objectIdFromString.oid).to(equal(oid))
         expect(objectIdFromString.timestamp).to(equal(timestamp))
     }
+
+    /// Test AnyBSONValue Hashable conformance
+    func testHashable() throws {
+        let expected = CodecTests.AllBSONTypes(
+                double: Double(2),
+                string: "hi",
+                doc: ["x": 1],
+                arr: [1, 2],
+                binary: try Binary(base64: "//8=", subtype: .generic),
+                oid: ObjectId(fromString: "507f1f77bcf86cd799439011"),
+                bool: true,
+                date: Date(timeIntervalSinceReferenceDate: 5000),
+                code: CodeWithScope(code: "hi", scope: ["x": 1]),
+                int: 1,
+                ts: Timestamp(timestamp: 1, inc: 2),
+                int32: 5,
+                int64: 6,
+                dec: Decimal128("1.2E+10"),
+                minkey: MinKey(),
+                maxkey: MaxKey(),
+                regex: RegularExpression(pattern: "^abc", options: "imx"))
+
+        let values = Mirror(reflecting: expected).children.map { child in AnyBSONValue(child.value as! BSONValue) }
```

Comment:
is there a reason to use mirror here rather than just making `expected` an `[BSONValue]` to begin with?

Reply:
I switched it to this after looking at Jason's tests, actually.

Using this struct guarantees at the compile time that we cover all `BSONValue` cases in this test. So if we add a new `BSONValue` in the future (and update `AllBSONTypes` accordingly), it will cause this test to stop compiling unless we also update it to use the new type. Using just a list wouldn't cause any errors if a new `BSONValue` were introduced, so we'd likely forget to update this test.

Reply:
that makes sense, though I wonder if we will remember to update that struct... any ideas for enforcing that?

Reply:
We could include a comment near the definition of the `BSONValue` protocol instructing us to do so. 
Alternatively, we could include a test that does some mirroring stuff that verifies the number of children in `AllBSONTypes` is equivalent to the number of cases in `BSONType`.

Reply:
Filed [SWIFT-324](https://jira.mongodb.org/browse/SWIFT-324)